### DISCUSSION TITLE
Allow automations to send events to client-side (Websockets?)

### DISCUSSION THREAD

**Discussion by chrisshaw** [Budibase/budibase#2608] [Ideas]
(Apologies if this is already possible.)

It would be awesome if there were some way to programmatically route people to another page. Essentially, do what's possible in the Navigate to Action.

## Why?

- I want to keep my user's Budibase session "in sync" with another external app.

Additionally:
- I want to power a first-time onboarding experience, based on a status in the database

## Use case

Let's say I'm using a messaging app to live chat with customers; we'll call it, oh I dunno, Crisp. Let's say I'm using Budibase as a homebrew CRM. A live chat agent has 2 windows open, one with the live chat app and one with our Budibase CRM. Whenever the live chat agent switches conversations, they then manually go searching for the record page in Budibase. :cry: 

Let's say I have a different app that captures events from live chat agents. When they switch the conversation they are in, that fires an event with data on which customer they are now actively talking to.

I wish my intermediate app could then POST a request to a Budibase Automation webhook. That POST body would include some data like the live chat agent's ID and the customer's ID in Crisp. The Automation would then run a few steps:

1. look up the matching User in Budibase
2. look up the customer record in an external data source.
3. if the User was active...
4. **Navigate to, e.g., `customers//{{ customer_id }}/detail`**

**Comment by mjashanks** [comment]
Hey Chris.

Thanks for this - totally understand your use case. 

The issue with what you describe is that automations are not connected to the user's web browser - they run completely on the server-side. So, whenever your automation fires, there is no way to communicate with the web app, to tell it to navigate to a different page.

This would be an awesome feature. However, I think it would require some big changes, including some way in the builder to handle "events" that get sent from automation to the user. It's a whole new concept that could be a minefield!

I think it's a lovely idea, but don't hold your breath :) it won't be happening anytime soon! We do review these ideas when defining the roadmap - so happy to keep this one around.

P.S. I renamed the discussion to reflect what would be required.

**Reply by chrisshaw** [reply]
Thanks for the thoughtful response. I also ran into a second use case: routing people to a first-time onboarding experience based on some status in the database. Without trying, I think something like this would be possible with conditional rendering, but (correct me if I'm wrong) depending on the entry point we'd have to repeat elements across pages.

While I do think it's most natural/powerful to have Automations integrate with the client, it's also possible the custom components feature that a number of people have been asking for would also allow me to cobble together a workaround, if I had access to the router.

**Reply by dncpax** [reply]
could we have a list of async actions with status like pending, executing, executed? the client app could poll this list...
